File: day01/zig/src/main.zig
const std = @import("std");

pub fn main() !void {
    var stdout_buf: [256]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buf);
    const stdout = &stdout_writer.interface;

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    var read_buf: [4096]u8 = undefined;
    var file_reader = file.reader(&read_buf);
    const reader = &file_reader.interface;

    var part1: i32 = 0;
    var part2: i32 = 0;
    var pos: i32 = 50;

    while (try reader.takeDelimiter('\n')) |raw_line| {
        const line = std.mem.trim(u8, raw_line, &std.ascii.whitespace);
        if (line.len == 0) continue;

        const dir = line[0];
        const dist = try std.fmt.parseInt(i32, line[1..], 10);

        const crossings: i32 = switch (dir) {
            'R' => @divTrunc(pos + dist, 100),
            'L' => blk: {
                if (pos == 0) {
                    break :blk @divTrunc(dist, 100);
                } else if (pos <= dist) {
                    break :blk @divTrunc(dist - pos, 100) + 1;
                } else {
                    break :blk 0;
                }
            },
            else => 0,
        };

        pos = switch (dir) {
            'R' => @mod(pos + dist, 100),
            'L' => @mod(pos - dist, 100),
            else => pos,
        };

        part2 += crossings;
        if (pos == 0) part1 += 1;
    }

    try stdout.print("Part 1: {d}\n", .{part1});
    try stdout.print("Part 2: {d}\n", .{part2});
    try stdout.flush();
}

File: day01/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let instructions: Vec<&str> = input.lines().map(|l| l.trim()).filter(|l| !l.is_empty()).collect();

    println!("Part 1: {}", count_zeros_landing(&instructions));
    println!("Part 2: {}", count_zeros_all_clicks(&instructions));
}

fn count_zeros_landing(instructions: &[&str]) -> i32 {
    let mut pos: i32 = 50;
    let mut count = 0;

    for instr in instructions {
        let dir = instr.chars().next().unwrap();
        let dist: i32 = instr[1..].parse().unwrap();

        pos = match dir {
            'R' => pos + dist,
            'L' => pos - dist,
            _ => pos,
        };
        pos = pos.rem_euclid(100);

        if pos == 0 {
            count += 1;
        }
    }
    count
}

fn count_zeros_all_clicks(instructions: &[&str]) -> i32 {
    let mut pos: i32 = 50;
    let mut count = 0;

    for instr in instructions {
        let dir = instr.chars().next().unwrap();
        let dist: i32 = instr[1..].parse().unwrap();

        let crossings = match dir {
            'R' => (pos + dist) / 100,
            'L' => {
                if pos == 0 {
                    dist / 100
                } else if pos <= dist {
                    (dist - pos) / 100 + 1
                } else {
                    0
                }
            }
            _ => 0,
        };

        pos = match dir {
            'R' => pos + dist,
            'L' => pos - dist,
            _ => pos,
        };
        pos = pos.rem_euclid(100);

        count += crossings;
    }
    count
}
